File: EMR_BMED.Backend/Controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Authorization;
using Microsoft.IdentityModel.Tokens;

using EMR_BMED.Backend.Services;
using EMR_BMED.Backend.Models;
using EMR_BMED.Backend.Exceptions;

namespace EMR_BMED.Backend.Controllers
{
  [ApiController]
  [Route("[controller]")]
  public class AuthController(AuthService authService) : ControllerBase
  {
    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginDto credentials)
    {
      try
      {
        string token = await authService.LoginAsync(credentials);
        return Ok(new { token });
      }
      catch (UserNotFoundException ex)
      {
        return NotFound(new { email = ex.Message });
      }
      catch (IncorrectPasswordException ex)
      {
        return Unauthorized(new { password = ex.Message });
      }
    }

    [HttpPost("register/patient")]
    public async Task<IActionResult> Register([FromBody] PatientModel patientData)
    {
      try
      {
        await authService.RegisterAsync(patientData);
        return Ok();
      }
      catch (EmailIsTakenException ex)
      {
        return Conflict(new { email = ex.Message });
      }
    }

    [HttpPost("register/doctor")]
    public async Task<IActionResult> Register([FromBody] DoctorModel doctorData)
    {
      try
      {
        await authService.RegisterAsync(doctorData);
        return Ok();
      }
      catch (EmailIsTakenException ex)
      {
        return Conflict(new { email = ex.Message });
      }
    }

    // Temporary
    [Authorize]
    [HttpGet("whoami")]
    public async Task<IActionResult> WhoAmI([FromHeader(Name = "Authorization")] string authHeader)
    {
      try
      {
        var user = await authService.WhoAmI(authHeader);
        return Ok(new { user.Name, user.Surname });
      }
      catch (UserNotFoundException)
      {
        return NotFound(new { username = $"Access token doesn't belong to a user" });
      }
      catch (SecurityTokenMalformedException ex)
      {
        return BadRequest(new { message = ex.Message });
      }
    }
  }
}

File: EMR_BMED.Backend.JS/controllers/auth.controller.js
import express from "express";
const router = express.Router();

import authService from "../services/auth.service.js";

router
  .post("/login", (req, res) => {
    try {
      const token = authService.login(req.body);

      return res.send({ token });
    } catch (err) {
      if (err.name === "UserNotFoundError") {
        return res.status(404).send({ email: err.message });
      } else if (err.name === "IncorrectPasswordError") {
        return res.status(401).send({ password: err.message });
      }
    }
  })
  .post("/register/patient", (req, res) => {
    try {
      authService.register(req.body);
      return res.status(200).send({});
    } catch (err) {
      if (err.name === "EmailIsTakenError") {
        return res.status(409).send({ email: err.message });
      }
    }
  })
  .post("/register/doctor", (req, res) => {
    try {
      authService.register(req.body);
      return res.status(200).send({});
    } catch (err) {
      if (err.name === "EmailIsTakenError") {
        return res.status(409).send({ email: err.message });
      }
    }
  })
  .get("/whoami", (req, res) => {
    try {
      const user = authService.whoAmI(req.headers.authorization);
      return res.send(user);
    } catch (err) {
      if (err.name === "UserNotFoundError") {
        return res.status(404).send({ message: err.message });
      } else {
        return res.status(400).send({ message: err.message });
      }
    }
  });

export default router;
